text/2019-03-19-calculated-commit-ts.md

```diff
@@ -0,0 +1,190 @@
+# Calculated Commit TS
+
+## Summary
+
+This RFC proposes a feature of calculating `commit_ts` instead of allocating
+`commit_ts` from PD to reduce latency of committing.
+
+## Motivation
+
+TiKV uses Percolator's transaction algorithm. When committing a transaction,
+there are a few steps to do:
+
+1. Prewrite all keys with the transaction's `start_ts`
+2. Allocate the `commit_ts` from PD.
+3. Commit the transaction with `commit_ts`.
+
+The `commit_ts` allocated from PD is important in providing Snapshot Isolation,
+however allocating it from PD is not the only possible way. It's possible to
+calculate it instead of allocating it for every transaction. Therefore, we can
+remove the latency of waiting for PD's response, thereby reducing commit
+latency.
+
+## Detailed design
+
+### How to calculate a valid `commit_ts`
+
+To provide Snapshot Isolation, we should ensure all transactional reads are
+repeatable. The `commit_ts` should be large enough so that the transaction will
+not be committed before a valid read. Otherwise, Repeatable Read will be broken.
+For example:
+
+1. Txn1 gets `start_ts` 100
+2. Txn2 gets `start_ts` 200
+3. Txn2 reads key `"k1"` and gets value `"1"`
+4. Txn1 prewrites `"k1"` with value `"2"`
+5. Txn1 commits with `commit_ts` 101
+6. Tnx2 reads key `"k1"` and gets value `"2"`
+
+Txn2 reads `"k1"` twice but gets two different results. If `commit_ts` is
+allocated from PD, this will not happen, because Txn2's first read must happen
+before Txn1's prewrite, and Txn1's `commit_ts` must be requested after finishing
+prewrite, which means, Txn2's `commit_ts` must be larger than Txn1's `start_ts`.
```

Comment:
```suggestion
prewrite. This means Txn2's `commit_ts` must be larger than Txn1's `start_ts`.
```

Reply:
I want the both two parts around the "which means" be in the `because` clause. How is it if I change it to this:
> If `commit_ts` is allocated from PD, this will not happen, because Txn2's first read must happen before Txn1's prewrite while Txn1's `commit_ts` must be requested after finishing prewrite, and as a result Txn2's `commit_ts` must be larger than Txn1's `start_ts`.

Reply:
The point is to avoid run-on sentences. Your re-writing is good. And we may want to split the sentence again as below:
> If commit_ts is allocated from PD, this will not happen, because Txn2's first read must happen before Txn1's prewrite while Txn1's commit_ts must be requested after finishing prewrite. And as a result, Txn2's commit_ts must be larger than Txn1's start_ts.

Reply:
Done